### ISSUE TITLE
Regression in implementing wp_cache_flush_runtime

### ISSUE THREAD

**Issue by rrennick** [Prospress/action-scheduler#960]
Ref: p6q8Tx-37p-7529-p2

`wp_cache_flush_runtime` was implemented in #827 with the intention to 

> WordPress 6.0.0 introduced wp_cache_flush_runtime() as a way to flush the runtime cache without impacting the external object cache.

The PHPDoc description on `wp_cache_flush_runtime` is

> Removes all cache items from the in-memory runtime cache.

Prior to #827, external caches were only flushed with `wp_cache_flush` when the `action_scheduler_queue_runner_flush_cache` returned `true`. [`wp_cache_flush_runtime` is an alias for `wp_cache_flush`](https://github.com/WordPress/wordpress-develop/blob/6.2/src/wp-includes/cache.php#L280-L282). I don't see any evidence in the current WP code that excludes external caches.

 With the current logic, a site on WP 6.2 will flush external cache without checking the filter. We should restore the original logic while adding support for `wp_cache_flush_runtime` which may be expanded in the future.

**Comment by barryhughes** [comment]
I'm not quite understanding the problem just yet 🤔 ... but let me talk through my high-level understanding of how this works, and then we can look at a few examples.

First, it's worth noting that in an environment with **no** external object cache, there is effectively no meaningful distinction between `wp_cache_flush` and `wp_cache_flush_runtime`: the runtime cache is the only object cache we have. So, by default, it makes sense that one is an alias of the other.

Things change when we add an external object cache. Generally speaking, we now have an external (persisted) object cache in *addition* to the runtime cache ... but not all caching implementations support the ability to flush only the runtime cache.

Let's look at a few examples, all using WordPress 6.2 and, for simplicity, assuming there are no filters for `action_scheduler_queue_runner_flush_cache`.

### No external cache

In this scenario, the [flushing_runtime_cache_explicitly_supported test](https://github.com/woocommerce/action-scheduler/blob/3.6.0/classes/ActionScheduler_QueueRunner.php#L193) passes, because WordPress core [supports flushing the runtime cache](https://github.com/WordPress/wordpress-develop/blob/6.2.2/src/wp-includes/cache.php#L320-L322), and we then go right ahead and [flush the runtime cache](https://github.com/woocommerce/action-scheduler/blob/trunk/classes/ActionScheduler_QueueRunner.php#L197). 

As you note, this is effectively the same as a direct call to [`wp_cache_flush`](https://github.com/woocommerce/action-scheduler/blob/3.6.0/classes/ActionScheduler_QueueRunner.php#L197). As I read it, the behavior here is exactly as it was in [3.4.2](https://github.com/woocommerce/action-scheduler/blob/3.4.2/classes/ActionScheduler_QueueRunner.php#L183-L187) (in the sense that, then as now, if there is no external object cache we go ahead and flush the cache).

### External cache with support for flushing the runtime cache

We could use [Redis Cache 2.4.1](https://wordpress.org/plugins/redis-cache/) if we need a concrete example. The logic is basically the same as before, in the sense that the [flushing_runtime_cache_explicitly_supported test](https://github.com/woocommerce/action-scheduler/blob/3.6.0/classes/ActionScheduler_QueueRunner.php#L193) passes, and we go ahead and [flush the runtime cache](https://github.com/woocommerce/action-scheduler/blob/trunk/classes/ActionScheduler_QueueRunner.php#L197). 

If we look inside the caching implementation, though, we can see that we have separation between runtime cache and persisted cache and, in this case, [only the runtime cache is cleared](https://plugins.trac.wordpress.org/browser/redis-cache/tags/2.4.1/includes/object-cache.php#L173), not the persisted cache.

### External cache *without* support for flushing the runtime cache

We could use [Memcached 4.0.0](https://wordpress.org/plugins/memcached/) as an example here. This time, the [flushing_runtime_cache_explicitly_supported test](https://github.com/woocommerce/action-scheduler/blob/3.6.0/classes/ActionScheduler_QueueRunner.php#L193) will **not** pass: that is because this version of the Memcached drop-in does not define its own version of `wp_cache_supports`, therefore [this fallback](https://github.com/WordPress/wordpress-develop/blob/6.2.2/src/wp-includes/cache-compat.php#L198-L200) (which always returns false) is used. 

We now enter [this branch](https://github.com/woocommerce/action-scheduler/blob/3.6.0/classes/ActionScheduler_QueueRunner.php#L194) which notices there is an external object cache in play, and so takes no further action.

**Comment by rrennick** [comment]
@barryhughes Thanks for the clarification. In WP 6.0, the second condition would be true for any persistent object cache that has not added `wp_cache_supports` support as the default function wasn't [added until 6.1](https://github.com/WordPress/wordpress-develop/blob/6.2/src/wp-includes/cache-compat.php#L188:L200). 

```
$flushing_runtime_cache_implicitly_supported = ! function_exists( 'wp_cache_supports' ) && function_exists( 'wp_cache_flush_runtime' );
```

**Comment by barryhughes** [comment]
Yep, so if we're running under WP 6.0 and the second condition is met, [we go ahead and call `wp_cache_flush_runtime()`](https://github.com/woocommerce/action-scheduler/blob/3.6.0/classes/ActionScheduler_QueueRunner.php#L197). Three things are possible from here:

- No external cache → the cache (which will be the runtime cache) is flushed anyway.
- External cache that supports flushing the runtime cache → runtime cache is flushed.
- External cache without support for flushing the runtime cache → [nothing happens](https://github.com/WordPress/wordpress-develop/blob/6.0.0/src/wp-includes/cache-compat.php#L141).

So that last case is the one you are concerned with, and what you are suggesting is we make it possible (via the existing filter) to flush the cache anyway?

**Comment by barryhughes** [comment]
✍🏼 Just to help others who might be following, and to try and keep the conversation accessible to all: p6q8Tx-37p-7529-p2 (referred to in the opening post) is a doc containing an observation that WooCommerce may be flushing the cache (not just the runtime cache) too often/too aggressively. Of course, WooCommerce includes Action Scheduler as one of its bundled dependencies, which is part of the reason we're having this discussion.

**Comment by rrennick** [comment]
@barryhughes I think the point of confusion is that I am looking at the main [`wp_cache_flush_runtime`](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/cache.php#L265) and you are looking at the [compat version](https://github.com/WordPress/wordpress-develop/blob/6.0.0/src/wp-includes/cache-compat.php#L141). 

The [developer docs](https://developer.wordpress.org/reference/functions/wp_cache_flush_runtime/) have no mention of the compat version.

**Comment by barryhughes** [comment]
Gotcha. 

As I understand it, `cache-compat.php` is always loaded (and conditionally defines functions such as `wp_cache_flush_runtime` if they don't already exist/if an external cache is present but does not define them). `cache.php` is *only* loaded if an external cache is not present.

https://github.com/WordPress/wordpress-develop/blob/6.2.2/src/wp-includes/load.php#L736-L740

**Comment by jeremysimmons** [comment]
@rrennick, @barryhughes 

Sorry to necro the ticket. I'll open a new one if you prefer

I ran into this today on a site in wpengine.
```
ErrorException: User Notice: Function wp_cache_flush_runtime was called <strong>incorrectly</strong>. Your object cache implementation does not support flushing the in-memory runtime cache. Please see <a href="https://wordpress.org/documentation/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 6.1.0.)
#13 /wp-includes/functions.php(5865): _doing_it_wrong
#12 /wp-includes/cache-compat.php(145): wp_cache_flush_runtime
#11 /wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_QueueRunner.php(191): ActionScheduler_QueueRunner::clear_caches
#10 /wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_QueueRunner.php(175): ActionScheduler_QueueRunner::do_batch
#9 /wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_QueueRunner.php(136): ActionScheduler_QueueRunner::run
#8 /wp-includes/class-wp-hook.php(308): WP_Hook::apply_filters
#7 /wp-includes/class-wp-hook.php(332): WP_Hook::do_action
#6 /wp-includes/plugin.php(517): do_action
#5 /wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_AsyncRequest_QueueRunner.php(52): ActionScheduler_AsyncRequest_QueueRunner::handle
#4 /wp-content/plugins/woocommerce/includes/libraries/wp-async-request.php(147): WP_Async_Request::maybe_handle
#3 /wp-includes/class-wp-hook.php(308): WP_Hook::apply_filters
#2 /wp-includes/class-wp-hook.php(332): WP_Hook::do_action
#1 /wp-includes/plugin.php(517): do_action
#0 /wp-admin/admin-ajax.php(188): null
```



The implementation of `wp_cache_flush_runtime` does the check `! wp_cache_supports( 'flush_runtime' )`, [source](https://github.com/WordPress/WordPress/blob/ca421dd3c0274be8956f9822640713c6cd4c640b/wp-includes/cache-compat.php#L140) 

Tt seems like ActionScheduler should be checking if the cache supports calling `wp_cache_flush_runtime` before calling it, to avoid the `_doing_it_wrong` error.

If you agree, I'll submit a PR.

**Comment by barryhughes** [comment]
Hi @jeremysimmons, 

The current WooCommerce release is packaged with [Action Scheduler 3.5.4](https://plugins.trac.wordpress.org/browser/woocommerce/tags/7.8.1/packages/action-scheduler/action-scheduler.php#L8) and, when that was released (at the start of the year), WordPress did not yet provide a way of checking if flushing the runtime cache was supported. That subsequently changed and, once we became aware of this, we in turn updated Action Scheduler. If you take a peek at the code in our [3.6.1](https://github.com/woocommerce/action-scheduler/tree/3.6.1) release, for instance, you will see we already perform the necessary check:

💻 [`ActionScheduler_QueueRunner::clear_caches()`](https://github.com/woocommerce/action-scheduler/blob/3.6.1/classes/ActionScheduler_QueueRunner.php#L193-L194)

That to say, this shouldn't be a problem if you are using the latest version of Action Scheduler—which we also expect to be included in the [next WooCommerce release](https://plugins.trac.wordpress.org/browser/woocommerce/tags/7.9.0-beta.2/packages/action-scheduler/action-scheduler.php#L8).

If you do still believe there is a problem in the latest version of Action Scheduler, though, please do create a bug report (and optionally a PR, if you have time), and we will be happy to take a look.

**Comment by jeremysimmons** [comment]
@barryhughes thanks for the reply! Thanks for sharing the extra history too. I hadn't researched that deeply...

>  this shouldn't be a problem if you are using the latest version of Action Scheduler

per the https://actionscheduler.org/faq/, 
> Action Scheduler is designed to be used and released in plugins. It avoids redeclaring public API functions when more than one copy of the library is being loaded by different plugins. It will also load only the most recent version of itself (by checking registered versions after all plugins are loaded on the 'plugins_loaded' hook).

I believe that means I can install action-scheduler as a stand-alone plugin and will it's version supersede the one bundled with Woocommerce. Is that right?

**Comment by barryhughes** [comment]
Exactly—if you install Action Scheduler as a standalone plugin, and so long as it is the latest version, then that ought to be the version that is loaded.

Of course, the reverse is also true: if your standalone copy of Action Scheduler becomes outdated, but a plugin like WooCommerce includes a more recent version, then it would be WooCommerce's copy that 'wins'.

tl;dr → the most up-to-date active version is the one that will be loaded.
